### PR TITLE
Ruby: Fixes for x64-mingw-ucrt (PR to master)

### PR DESCRIPTION
**Pull request by johnnyshields**
Follow-up to #29684 @apolcyn @gnossen 

#### TODOS

- [x] Add artifact target for x64-mingw-ucrt
- [x] Fix `extconf.rb` to correctly pickup the UCRT grpc build. Adds a `CPP_FLAG` called `GRPC_WINDOWS_UCRT`
- [x] Upgrade rake_compiler_dock to get Ruby [shared object fix](https://github.com/rake-compiler/rake-compiler-dock/issues/69)

In addition, I've renamed intermediate build files for clarity. These are renamed to `grpc_c.so` during the build process, so should not affect anything in the output.

```
grpc_c.32.ruby --> grpc_c.32-msvcrt.ruby
grpc_c.64.ruby --> grpc_c.64-msvcrt.ruby
```

### CODE REVIEW

**Review by apolcyn — APPROVED** [review]

**Comment by apolcyn** [comment]
Hmm, this new windows library is failing to build:

```
linking shared-object grpc/grpc_c.so
/usr/bin/x86_64-w64-mingw32-ld: cannot find -lx64-ucrt-ruby310
collect2: error: ld returned 1 exit status
make: *** [Makefile:265: grpc_c.so] Error 1
rake aborted!
```

**Comment by johnnyshields** [comment]
@flavorjones any bandwidth to help on this?

**Comment by johnnyshields** [comment]
The error above indicates that the shared object `x64-ucrt-ruby310.so` can't be found, thus the linker command `x86_64-w64-mingw32-ld` fails.

On non-UCRT builds (MSVCRT), this file is called `x64-msvcrt-ruby300.so`. Perhaps Ruby 3.1.0 is either (a) failing to build or (b) outputting the `msvcrt` named file instead.

**Comment by johnnyshields** [comment]
@flavorjones @larskanis are you aware of any reason the linker command `x86_64-w64-mingw32-ld` wouldn't pick up the `-lx64-ucrt-ruby310` arg on Windows 3.1.0 UCRT?

It seems to be working fine on the older MSVCRT Docker images (`-lx64-msvcrt-ruby300` etc)

**Comment by johnnyshields** [comment]
@apolcyn I was able to get this to build using the following hacks:

1. Jury-rig the Makefile to copy `x64-ucrt-ruby310.dll` from a Windows Ruby installation to the build dir, and rename it to `libx64-ucrt-ruby310.a`

```
(cp #{File.join(grpc_root, 'x64-ucrt-ruby310.dll')} #{File.join(output_dir, 'libx64-ucrt-ruby310.a')})
```

2. Modify this line to `grpc_c.64-ucrt.ruby`: https://github.com/grpc/grpc/blob/master/src/ruby/ext/grpc/rb_loader.c#L26

Item (1) indicates that the issue is that the linker command isn't picking up the Ruby shared object like it should. Item (2) we'll need to pass-in an environment variable like `GRP_WINDOWS_UCRT` here.

**Comment by flavorjones** [comment]
@johnnyshields Sorry I missed this thread, I responded at https://github.com/rake-compiler/rake-compiler-dock/issues/69 with a patch that addresses this issue in a slightly different way:

```patch
commit 3e22505 (HEAD -> flavorjones-v1.46.x-backport-ruby-x64-mingw-ucrt)
Author: Mike Dalessio <mike.dalessio@gmail.com>
Date:   2022-06-23 15:29:53 -0400

    fix: do not statically link the ruby windows ucrt library
    
    The error being raised by the linked when trying to create the
    grpc_c.so was:
    
    > linking shared-object grpc/grpc_c.so
    > /usr/bin/x86_64-w64-mingw32-ld: cannot find -lx64-ucrt-ruby310
    > collect2: error: ld returned 1 exit status
    > make: *** [Makefile:265: grpc_c.so] Error 1
    
    as described at
    https://github.com/rake-compiler/rake-compiler-dock/issues/69

diff --git a/src/ruby/ext/grpc/extconf.rb b/src/ruby/ext/grpc/extconf.rb
index e0ca9b7..844f817 100644
--- a/src/ruby/ext/grpc/extconf.rb
+++ b/src/ruby/ext/grpc/extconf.rb
@@ -16,6 +16,7 @@
 require 'mkmf'
 
 windows = RUBY_PLATFORM =~ /mingw|mswin/
+windows_ucrt = RUBY_PLATFORM =~ /(mingw|mswin).*ucrt/
 bsd = RUBY_PLATFORM =~ /bsd/
 darwin = RUBY_PLATFORM =~ /darwin/
 linux = RUBY_PLATFORM =~ /linux/
@@ -100,7 +101,7 @@
 
 $LDFLAGS << ' -Wl,-wrap,memcpy' if linux
 $LDFLAGS << ' -static-libgcc -static-libstdc++' if linux
-$LDFLAGS << ' -static' if windows
+$LDFLAGS << ' -static' if windows && !windows_ucrt
 
 $CFLAGS << ' -std=c99 '
 $CFLAGS << ' -Wall '
```

I poked @larskanis on that issue, this may be a file naming issue in the rake-compiler-dock image.

**Comment by flavorjones** [comment]
See https://github.com/rake-compiler/rake-compiler-dock/issues/69#issuecomment-1164971004 for the underlying diagnosis. My hope is that we can cut a new version of rake-compiler-dock shortly to fix this.

**Comment by flavorjones** [comment]
Proposed upstream fix at https://github.com/rake-compiler/rake-compiler-dock/pull/70

**Comment by johnnyshields** [comment]
Update: @flavorjones 's PR is merged, guess we're still waiting on the official release + new Docker image.

**Comment by johnnyshields** [comment]
@apolcyn this is ready for final test and review. Note that all Ruby Rake Compiler Docker images are going to need to be re-published (not just `x64-mingw-ucrt` I think.)

```
Unable to find image 'grpctesting/rake_x64-mingw-ucrt:0b8bed3e125d8952ccc8938f905d9792c84077f5' locally
docker: Error response from daemon: manifest for grpctesting/rake_x64-mingw-ucrt:0b8bed3e125d8952ccc8938f905d9792c84077f5 not found: manifest unknown: manifest unknown.
```

**Comment by apolcyn** [comment]
I've pushed dockerfile changes to CI. Re-running CI

**Comment by johnnyshields** [comment]
@apolcyn ran the build and everything worked successfully. Please try it on your end to verify.

**Comment by apolcyn** [comment]
I think this had merge conflicts with https://github.com/grpc/grpc/pull/27660, PTAL

**Comment by johnnyshields** [comment]
@apolcyn conflict resolved

**Comment by johnnyshields** [comment]
(Just squashed commits and fixed EasyCLA)

File: src/ruby/ext/grpc/extconf.rb
Comment (apolcyn) [inline src/ruby/ext/grpc/extconf.rb:89]: minor naming nit: this isn't a typical "GPR" macro, like the rest of the ones defined [here](https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/port_platform.h) for the C-core library, so it feels out of place to use that here.

Can we rename this to `GRPC_RUBY_WINDOWS_UCRT` ?

**Review by apolcyn — APPROVED** [review]
approving with a minor comment.

@veblush, can I get a second review?

File: src/ruby/ext/grpc/extconf.rb
Comment (johnnyshields) [reply src/ruby/ext/grpc/extconf.rb]: @apolcyn done

**Comment by johnnyshields** [comment]
@veblush can you please help review this PR?

**Comment by apolcyn** [comment]
tests failures unrelated

**Comment by apolcyn** [comment]
@johnnyshields , sorry this didn't make it into 1.47 or 1.48.

The 1.49 branch cut is scheduled for 2-3 weeks, at which point there will be a pre-release containing this PR.

We also have nightly packages from the head of master branch in https://packages.grpc.io/. For example see the most recent ones in nightly builds in https://packages.grpc.io/archive/2022/07/93fbacb5343ea8cf993f779cee0847aed107bdc7-81b4ae53-8f32-463e-b5c6-b026f8b12cf5/index.xml, which has [grpc-1.49.0.dev-x64-mingw-ucrt.gem](https://packages.grpc.io/archive/2022/07/93fbacb5343ea8cf993f779cee0847aed107bdc7-81b4ae53-8f32-463e-b5c6-b026f8b12cf5/ruby/grpc-1.49.0.dev-x64-mingw-ucrt.gem).

Are those nightly packages sufficient for you, until 1.49 is cut?

**Comment by johnnyshields** [comment]
@apolcyn no major rush, just want to see it released so I can check the box on my end. Totally fine to get it in 1.49 in 2-3 weeks, please just take care that Ruby gem is pushed at that time (as it seems some versions don't get a gem release...)

**Comment by apolcyn** [comment]
FYI, this package was just pushed as a part of the `1.49.0.pre1` RC

**Pull request merged**
